### PR TITLE
Implement outer for element-wise functions

### PR DESCRIPTION
**Pull request by npolina4**
Implemented outer for binary element-wise functions.
Updated dpnp.outer by using new ufunc outer (dpnp.multiply.outer)
Added tests for different shape and tests compare for compare with dpnp.outer.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

**Review by antonwolfy — APPROVED** [review]

**Review by antonwolfy — COMMENTED** [review]
I have only one general comments that we missed docstring documentation for class of universal function (description of attributes like `nin` and methods like `outer`).
But it can be addressed by a separate PR.

**Review by antonwolfy — COMMENTED** [review]
It's quite a mix of updating different functionality in one PR.
It would be much more better to split into 3 parts by affecting functionality, like updates in:

1. dpnp.flatten
3. ufunc.outer
2. dpnp.outer

File: dpnp/dpnp_iface_linearalgebra.py [hunk 1]
```diff
@@ -851,19 +848,25 @@ def matmul(
...
+        Second input vector. Input is flattened if not already 1-dimensional.
+    out(M, N) : {dpnp.ndarray, usm_ndarray}, optional
+        A location where the result is stored
+
+    Returns
+    -------
+    out(M, N) : dpnp.ndarray
+        out[i, j] = a[i] * b[j]
 
     See Also
```
[the remark this answers is not in the document]
Comment (npolina4) [reply dpnp/dpnp_iface_linearalgebra.py:871]: We do not have special link to ufunc.outer yet. Will be added after updating docs for ufunc.

File: dpnp/dpnp_algo/dpnp_elementwise_common.py [hunk 2]
```diff
@@ -341,6 +341,105 @@ def __call__(
...
+            First input array.
+        x2 : {dpnp.ndarray, usm_ndarray}
+            Second input array.
+        out : {None, dpnp.ndarray, usm_ndarray}, optional
+            Output array to populate.
+            Array must have the correct shape and the expected data type.
+        order : {"C", "F", "A", "K"}, optional
+            Memory layout of the newly output array, Cannot be provided
+            together with `out`. Default: "K".
+        dtype : dtype, optional
```
Comment (antonwolfy) [inline dpnp/dpnp_algo/dpnp_elementwise_common.py:370]:
```suggestion
        dtype : {None, dtype}, optional
```

File: dpnp/dpnp_algo/dpnp_elementwise_common.py [hunk 3]
```diff
@@ -341,6 +341,105 @@ def __call__(
...
+        Parameters
+        ----------
+        x1 : {dpnp.ndarray, usm_ndarray}
+            First input array.
+        x2 : {dpnp.ndarray, usm_ndarray}
+            Second input array.
+        out : {None, dpnp.ndarray, usm_ndarray}, optional
+            Output array to populate.
+            Array must have the correct shape and the expected data type.
+        order : {"C", "F", "A", "K"}, optional
```
Comment (antonwolfy) [inline dpnp/dpnp_algo/dpnp_elementwise_common.py:367]: numpy support `order = None` also
```suggestion
        order : {None, "C", "F", "A", "K"}, optional
```

File: dpnp/dpnp_algo/dpnp_elementwise_common.py [hunk 4]
```diff
@@ -341,6 +341,105 @@ def __call__(
...
+                            for compatibility with old code.
+
+        :obj:`dpnp.tensordot` : dpnp.tensordot(a, b, axes=((), ())) and
+                                dpnp.multiply.outer(a, b) behave same for all
+                                dimensions of a and b.
+
+        Examples
+        --------
+        >>> import dpnp as np
+        >>> np.multiply.outer([1, 2, 3], [4, 5, 6])
```
Comment (antonwolfy) [inline dpnp/dpnp_algo/dpnp_elementwise_common.py:399]: That doesn't work and not intended to be supported:
```python
np.multiply.outer([1, 2, 3], [4, 5, 6])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 np.multiply.outer([1, 2, 3], [4, 5, 6])

File ~/code/dpnp/dpnp/dpnp_algo/dpnp_elementwise_common.py:423, in DPNPBinaryFunc.outer(self, x1, x2, out, where, order, dtype, subok, **kwargs)
    344 def outer(
    345     self,
    346     x1,
   (...)
    353     **kwargs,
    354 ):
    355     """
    356     Apply the ufunc op to all pairs (a, b) with a in A and b in B.
    357
   (...)
    420
    421     """
--> 423     dpnp.check_supported_arrays_type(
    424         x1, x2, scalar_type=True, all_scalars=False
    425     )
    426     if dpnp.isscalar(x1) or dpnp.isscalar(x2):
    427         _x1 = x1

File ~/code/dpnp/dpnp/dpnp_iface.py:330, in check_supported_arrays_type(scalar_type, all_scalars, *arrays)
    327     if scalar_type and dpnp.isscalar(a):
    328         continue
--> 330     raise TypeError(
    331         f"An array must be any of supported type, but got {type(a)}"
    332     )
    334 if len(arrays) > 0 and not (all_scalars or any_is_array):
    335     raise TypeError(
    336         "At least one input must be of supported array type, "
    337         "but got all scalars."
    338     )

TypeError: An array must be any of supported type, but got <class 'list'>
```

File: dpnp/dpnp_algo/dpnp_elementwise_common.py [hunk 5]
```diff
@@ -341,6 +341,105 @@ def __call__(
...
+        >>> B = np.array([[1, 2, 3, 4]])
+        >>> B.shape
+        (1, 4)
+        >>> C = np.multiply.outer(A, B)
+        >>> C.shape; C
+        (2, 3, 1, 4)
+        array([[[[ 1,  2,  3,  4]],
+                [[ 2,  4,  6,  8]],
+                [[ 3,  6,  9, 12]]],
+            [[[ 4,  8, 12, 16]],
```
Comment (antonwolfy) [inline dpnp/dpnp_algo/dpnp_elementwise_common.py:417]:
```suggestion
              [[[ 4,  8, 12, 16]],
```

File: dpnp/dpnp_iface_linearalgebra.py [hunk 6]
```diff
@@ -851,62 +847,58 @@ def matmul(
...
-        `x1` and `x2` can not be scalars at the same time. Otherwise
-        the functions will be executed sequentially on CPU.
-        Input array data types are limited by supported DPNP :ref:`Data types`.
+    Parameters
+    ----------
+    a : {dpnp.ndarray, usm_ndarray}
+        First input vector. Input is flattened if not already 1-dimensional.
+    b : {dpnp.ndarray, usm_ndarray}
+        Second input vector. Input is flattened if not already 1-dimensional.
+    out : {dpnp.ndarray, usm_ndarray}, optional
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_linearalgebra.py:862]:
```suggestion
    out : {None, dpnp.ndarray, usm_ndarray}, optional
```

File: dpnp/dpnp_algo/dpnp_elementwise_common.py [hunk 7]
```diff
@@ -341,6 +341,107 @@ def __call__(
...
+        x1 : {dpnp.ndarray, usm_ndarray}
+            First input array.
+        x2 : {dpnp.ndarray, usm_ndarray}
+            Second input array.
+        out : {None, dpnp.ndarray, usm_ndarray}, optional
+            Output array to populate.
+            Array must have the correct shape and the expected data type.
+        order : {None, "C", "F", "A", "K"}, optional
+            Memory layout of the newly output array, Cannot be provided
+            together with `out`. Default: "K".
```
Comment (antonwolfy) [inline dpnp/dpnp_algo/dpnp_elementwise_common.py:369]:
```suggestion
            together with `out`. Default: ``"K"``.
```

**Review by antonwolfy — APPROVED** [review]
Thank you @npolina4 No more comment from me.
Adding documentation page for `ufunc` should be handled by a separate PR,

**Pull request merged**